language/tools/move-package/src/resolution/resolving_table.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, collections::BTreeMap};
+
+use anyhow::{anyhow, Result};
+use move_core_types::account_address::AccountAddress;
+
+use crate::source_package::parsed_manifest::{NamedAddress, PackageName};
+
+/// A named address qualified with the package name whose scope it belongs to.
+pub type QualifiedAddress = (PackageName, NamedAddress);
+
+/// A data structure for unifying named addresses across packages according to renamings and
+/// assigning them numerical addresses.
+#[derive(Debug)]
+pub struct ResolvingTable {
+    /// Disjoint set data structure for assignments which can either hold no value, a fixed value or
+    /// a forwarding reference to another element in the table.  Each entry in the `redirection`
+    /// table gets a slot in the `assignment` table.
+    assignments: Vec<Assignment>,
+
+    /// Mapping named addresses to an entry in the `assignments` table.
+    redirection: BTreeMap<QualifiedAddress, usize>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Assignment {
+    Assign(Option<AccountAddress>),
+    Linked(usize),
```

Comment:
IIRC correctly, this was previously done with pointers (`Ref` or something similar)? Just checking to make sure I'm following 

Regardless, +1 to this sort of flattened design. Maybe it's too much time in type systems, but I always find these sort of custom index solutions a lot less error prone :)

Reply:
Indeed, this used to be done with an `Rc<RefCell<Option<AccountAddress>>>`